sdk/storage/azure-storage-common/inc/azure/storage/common/paged_response.hpp

```diff
@@ -0,0 +1,47 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#pragma once
+
+#include <cstdlib>
+#include <string>
+
+#include <azure/core/context.hpp>
+#include <azure/core/http/raw_response.hpp>
+
+namespace Azure {
+
+template <class Derived> class PagedResponse {
+public:
+  std::string CurrentPageToken;
+  std::string NextPageToken;
+  std::unique_ptr<Azure::Core::Http::RawResponse> RawResponse;
+
+  bool HasMorePages() const { return m_hasMorePages; }
+
+  void MoveToNextPage(const Azure::Core::Context& context = Azure::Core::Context())
+  {
+    static_assert(
+        std::is_base_of<PagedResponse, Derived>::value,
+        "The template argument \"Derived\" should derive from PagedResponse<Derived>.");
+
+    if (NextPageToken.empty())
+    {
+      m_hasMorePages = false;
+      return;
+    }
```

Comment:
Repeated calls to `MoveToNextPage()` after `HasMorePages()` returns false, should either no-op OR std::abort/throw.

So, maybe we should have a check like this:
```suggestion
  void MoveToNextPage(const Azure::Core::Context& context = Azure::Core::Context())
  {
    static_assert(
        std::is_base_of<PagedResponse, Derived>::value,
        "The template argument \"Derived\" should derive from PagedResponse<Derived>.");

    if (!m_hasMorePages || NextPageToken.empty())
    {
      m_hasMorePages = false;
      return;
    }
```